File: desktop/src/state/digestIssues.ts
import type { Entity } from '../hub/types';

/// Reader for the digest's structural-issues aggregation (transcript P5 §2 —
/// hub/internal/server/digest_issues.go). The hub folds the findings once and
/// serves them to both clients; this turns the wire map into the shape the
/// Issues drawer renders, and nothing more.
///
/// Pure and i18n-free on purpose: class keys stay keys, the component resolves
/// the display strings. That keeps this unit-testable (node --test, which CI
/// does NOT run — see CLAUDE.md) and keeps the ordering + seek-anchor rules in
/// ONE place instead of re-derived at each call site. Both are the shape of bug
/// this whole wedge exists to catch.

export type IssueSeverity = 'error' | 'warning' | 'info';

/// Worst-first. Must match the hub's issueSeverityRank (digest_issues.go) and
/// mobile's ordering — same severity order on both clients is a review anchor
/// of the plan, this family having shipped four mobile↔desktop misses already.
const SEVERITY_RANK: Record<string, number> = { error: 3, warning: 2, info: 1 };

export interface IssueSample {
  /// The transcript coordinate to seek to: the dense session_ordinal when the
  /// agent has one, else the per-agent seq. Same rule the turns list uses
  /// (AgentTranscript's turn-row anchor), resolved here so a caller never picks
  /// the wrong one — seq collides across a resumed session's agents, ordinal
  /// does not (ADR-042).
  coord?: number;
  ts?: string;
  /// Short headline for the row — the failing tool's name, the stop reason, the
  /// kind.key that changed shape. Absent when the class label carries it all.
  label?: string;
}

export interface IssueClass {
  cls: string;
  severity: IssueSeverity;
  count: number;
  samples: IssueSample[];
  /// True when the sample list is a prefix of `count` findings (the hub caps
  /// them at maxDigestErrorSeqs). Surfaced in the UI — a truncated list must
  /// never read as a complete one.
  capped: boolean;
}

export interface IssueSummary {
  total: number;
  worst?: IssueSeverity;
  classes: IssueClass[];
}

function severityOf(v: unknown): IssueSeverity {
  return v === 'error' || v === 'warning' || v === 'info' ? v : 'warning';
}

function numAt(arr: unknown, i: number): number | undefined {
  if (!Array.isArray(arr)) return undefined;
  const v = arr[i];
  return typeof v === 'number' && Number.isFinite(v) ? v : undefined;
}

function strAt(arr: unknown, i: number): string | undefined {
  if (!Array.isArray(arr)) return undefined;
  const v = arr[i];
  return typeof v === 'string' && v !== '' ? v : undefined;
}

/// Builds the drawer model from a digest entity. A hub that predates the
/// aggregation has no `issues` key at all and yields an empty summary, so the
/// drawer simply never offers itself — rather than rendering a clean bill of
/// health for a hub that never ran the checks.
export function readDigestIssues(digest: Entity | undefined): IssueSummary {
  const raw = digest?.['issues'];
  if (raw === null || raw === undefined || typeof raw !== 'object' || Array.isArray(raw)) {
    return { total: 0, classes: [] };
  }
  const classes: IssueClass[] = [];
  for (const [cls, entry] of Object.entries(raw as Record<string, unknown>)) {
    if (entry === null || typeof entry !== 'object' || Array.isArray(entry)) continue;
    const e = entry as Entity;
    const count = typeof e['count'] === 'number' && Number.isFinite(e['count']) ? (e['count'] as number) : 0;
    if (count <= 0) continue;
    const seqs = e['sample_seqs'];
    const ords = e['sample_ordinals'];
    const n = Array.isArray(seqs) ? seqs.length : 0;
    const samples: IssueSample[] = [];
    for (let i = 0; i < n; i += 1) {
      const ord = numAt(ords, i);
      const seq = numAt(seqs, i);
      samples.push({
        coord: ord !== undefined && ord > 0 ? ord : seq,
        ts: strAt(e['sample_ts'], i),
        label: strAt(e['sample_labels'], i),
      });
    }
    classes.push({ cls, severity: severityOf(e['severity']), count, samples, capped: count > samples.length });
  }

  // Severity-first, then the loudest class, then alphabetical so the order is
  // stable across renders and matches mobile.
  classes.sort((a, b) => {
    const bySev = (SEVERITY_RANK[b.severity] ?? 0) - (SEVERITY_RANK[a.severity] ?? 0);
    if (bySev !== 0) return bySev;
    if (b.count !== a.count) return b.count - a.count;
    return a.cls < b.cls ? -1 : a.cls > b.cls ? 1 : 0;
  });

  const total = classes.reduce((sum, c) => sum + c.count, 0);
  // Prefer the hub's own rollup: it ranks with the server's table, so a severity
  // this client doesn't know yet still tints correctly. Fall back to the
  // computed worst for a hub that predates the field.
  const wire = digest?.['issue_worst_severity'];
  const worst =
    typeof wire === 'string' && wire !== ''
      ? severityOf(wire)
      : classes.length > 0
        ? classes[0].severity
        : undefined;
  return { total, worst, classes };
}

File: lib/widgets/transcript/digest_issues.dart
/// Reader for the digest's structural-issues aggregation (transcript P5 §2 —
/// `hub/internal/server/digest_issues.go`). The hub folds the findings once and
/// serves them to both clients; this turns the wire map into the shape the
/// Issues sheet renders, and nothing more.
///
/// Pure and l10n-free on purpose: class keys stay keys, the widget resolves the
/// display strings. That makes it directly unit-testable — which matters more
/// than usual here, because this file's whole job is to hold the two rules that
/// must NOT be re-derived per call site:
///
///  * the severity ordering, which must match the hub's `issueSeverityRank` and
///    the desktop reader's (`desktop/src/state/digestIssues.ts`);
///  * the seek anchor — `sample_ordinals[i]` when > 0, else `sample_seqs[i]`.
///    `seq` COLLIDES across a resumed session's agents (ADR-042), so picking the
///    wrong one lands on the wrong row in exactly the case the sheet is most
///    useful. Same rule `_firstErrorSeq` and the turn rows already follow.
///
/// This family has shipped four mobile↔desktop misses (`callToolIdOf` ×3, the
/// stats strip #375); a shared, tested shape on each side is the answer.
library;

/// Severity ranks, worst first. Pinned to the hub's table.
const Map<String, int> _severityRank = {
  'error': 3,
  'warning': 2,
  'info': 1,
};

/// One sampled finding: where to seek, when it happened, what to call it.
class IssueSample {
  /// Transcript coordinate — the dense `session_ordinal` when the agent has
  /// one, else the per-agent `seq`. 0 when neither is usable.
  final int coord;
  final String ts;

  /// Short headline — the failing tool's name, the stop reason, the
  /// `kind.key` that changed shape. Empty when the class label carries it all.
  final String label;

  const IssueSample({required this.coord, this.ts = '', this.label = ''});
}

/// One issue class with its samples.
class IssueClassGroup {
  final String cls;
  final String severity;
  final int count;
  final List<IssueSample> samples;

  const IssueClassGroup({
    required this.cls,
    required this.severity,
    required this.count,
    required this.samples,
  });

  /// True when the sample list is a prefix of `count` findings (the hub caps
  /// them at `maxDigestErrorSeqs`). Surfaced in the sheet — a truncated list
  /// must never read as a complete one.
  bool get capped => count > samples.length;

  int get rank => _severityRank[severity] ?? 0;
}

/// The whole-run rollup the stat chip and the sheet read.
class IssueSummary {
  final int total;

  /// Worst severity present, or '' when there are none.
  final String worst;

  /// Classes ordered severity-first, then loudest, then alphabetical.
  final List<IssueClassGroup> classes;

  const IssueSummary({
    required this.total,
    required this.worst,
    required this.classes,
  });

  bool get isEmpty => classes.isEmpty;
}

int _asInt(dynamic v) {
  if (v is int) return v;
  if (v is double) return v.toInt();
  if (v is String) return int.tryParse(v) ?? 0;
  return 0;
}

String _asString(dynamic v) => v == null ? '' : v.toString();

String _severityOf(dynamic v) {
  final s = _asString(v);
  // An unknown severity degrades to the middle rather than dropping the row —
  // a rule added hub-side must never vanish from the client that has not
  // learned its name yet.
  return _severityRank.containsKey(s) ? s : 'warning';
}

/// Reads `issues` / `issue_worst_severity` off a digest body.
///
/// A hub that predates the aggregation carries no `issues` key at all and
/// yields an empty summary, so the sheet simply never offers itself — rather
/// than reporting a clean bill of health for a hub that never ran the checks.
IssueSummary readDigestIssues(Map<String, dynamic>? digest) {
  final raw = digest?['issues'];
  if (raw is! Map) return const IssueSummary(total: 0, worst: '', classes: []);

  final classes = <IssueClassGroup>[];
  for (final entry in raw.entries) {
    final v = entry.value;
    if (v is! Map) continue;
    final count = _asInt(v['count']);
    if (count <= 0) continue;
    final seqs = v['sample_seqs'];
    final ords = v['sample_ordinals'];
    final tss = v['sample_ts'];
    final labels = v['sample_labels'];
    final samples = <IssueSample>[];
    if (seqs is List) {
      for (var i = 0; i < seqs.length; i++) {
        final ord = (ords is List && i < ords.length) ? _asInt(ords[i]) : 0;
        final seq = _asInt(seqs[i]);
        samples.add(IssueSample(
          coord: ord > 0 ? ord : seq,
          ts: (tss is List && i < tss.length) ? _asString(tss[i]) : '',
          label: (labels is List && i < labels.length) ? _asString(labels[i]) : '',
        ));
      }
    }
    classes.add(IssueClassGroup(
      cls: entry.key.toString(),
      severity: _severityOf(v['severity']),
      count: count,
      samples: samples,
    ));
  }

  classes.sort((a, b) {
    final bySev = b.rank.compareTo(a.rank);
    if (bySev != 0) return bySev;
    final byCount = b.count.compareTo(a.count);
    if (byCount != 0) return byCount;
    return a.cls.compareTo(b.cls);
  });

  var total = 0;
  for (final c in classes) {
    total += c.count;
  }
  // Prefer the hub's own rollup: it ranks with the server's table, so a
  // severity this client cannot rank yet still tints correctly. Fall back to
  // the computed worst for a hub that predates the field.
  final wire = _asString(digest?['issue_worst_severity']);
  final worst = wire.isNotEmpty
      ? _severityOf(wire)
      : (classes.isEmpty ? '' : classes.first.severity);
  return IssueSummary(total: total, worst: worst, classes: classes);
}
